### PR TITLE
Refactor course cloning

### PR DESCRIPTION
**Pull request by Aaron2000119**
Part of https://github.com/Gamify-IT/issues/issues/387.

Cloning is now handled by the minigame-backends.

To test, start the docker-compose-complete(-dev), go to the lecturer-interface and clone a course.

Please also review:
similar:
https://github.com/Gamify-IT/crosswordpuzzle-backend/pull/32
https://github.com/Gamify-IT/finitequiz-backend/pull/19
https://github.com/Gamify-IT/chickenshock-backend/pull/38
https://github.com/Gamify-IT/towercrush-backend/pull/19

different:
https://github.com/Gamify-IT/bugfinder-backend/pull/9

### CODE REVIEW

File: src/main/java/de/unistuttgart/overworldbackend/service/AchievementStatisticService.java [hunk 1]
```diff
@@ -16,12 +16,20 @@
...
+    public void PlayerRepository(final PlayerRepository playerRepository) {
+        this.playerRepository = playerRepository;
+    }
+
     private AchievementStatisticRepository achievementStatisticRepository;
 
+    @Autowired
+    public void AchievementStatisticRepository(final AchievementStatisticRepository achievementStatisticRepository) {
+        this.achievementStatisticRepository = achievementStatisticRepository;
+    }
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/AchievementStatisticService.java]: What's the difference to previously?

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 2]
```diff
@@ -359,26 +373,15 @@ private MinigameTask cloneBugfinder(
...
                 return new MinigameTask(
                     Minigame.BUGFINDER,
                     minigameTask.getDescription(),
-                    config.getId(),
+                    cloneId,
                     minigameTask.getIndex()
                 );
             } catch (final FeignException e) {
-                log.debug(CLONE_ERROR_MESSAGE, e);
+                log.info(CLONE_ERROR_MESSAGE, e);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]: Erm… I think we need to have a discussion about appropriate log levels…
```suggestion
                log.error(CLONE_ERROR_MESSAGE, e);
```

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 3]
```diff
@@ -399,21 +402,15 @@ private MinigameTask cloneCrosswordpuzzle(
...
                 return new MinigameTask(
                     Minigame.CROSSWORDPUZZLE,
                     minigameTask.getDescription(),
-                    config.getId(),
+                    cloneId,
                     minigameTask.getIndex()
                 );
             } catch (final FeignException e) {
-                log.debug(CLONE_ERROR_MESSAGE, e);
+                log.info(CLONE_ERROR_MESSAGE, e);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]:
```suggestion
                log.error(CLONE_ERROR_MESSAGE, e);
```

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 4]
```diff
@@ -429,21 +426,15 @@ private MinigameTask cloneFinitequiz(
...
                 return new MinigameTask(
                     Minigame.FINITEQUIZ,
                     minigameTask.getDescription(),
-                    config.getId(),
+                    cloneId,
                     minigameTask.getIndex()
                 );
             } catch (final FeignException e) {
-                log.debug(CLONE_ERROR_MESSAGE, e);
+                log.info(CLONE_ERROR_MESSAGE, e);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]:
```suggestion
                log.error(CLONE_ERROR_MESSAGE, e);
```

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 5]
```diff
@@ -453,27 +444,21 @@ private MinigameTask cloneFinitequiz(
...
                 return new MinigameTask(
                     Minigame.TOWERCRUSH,
                     minigameTask.getDescription(),
-                    config.getId(),
+                    cloneId,
                     minigameTask.getIndex()
                 );
             } catch (final FeignException e) {
-                log.debug(CLONE_ERROR_MESSAGE, e);
+                log.info(CLONE_ERROR_MESSAGE, e);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]:
```suggestion
                log.error(CLONE_ERROR_MESSAGE, e);
```

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 6]
```diff
@@ -494,21 +479,15 @@ private MinigameTask cloneChickenshock(
...
                 return new MinigameTask(
                     Minigame.CHICKENSHOCK,
                     minigameTask.getDescription(),
-                    config.getId(),
+                    cloneId,
                     minigameTask.getIndex()
                 );
             } catch (final FeignException e) {
-                log.debug(CLONE_ERROR_MESSAGE, e);
+                log.info(CLONE_ERROR_MESSAGE, e);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]:
```suggestion
                log.error(CLONE_ERROR_MESSAGE, e);
```

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 7]
```diff
@@ -39,18 +37,34 @@ public class CourseService {
...
+
     CrosswordpuzzleClient crosswordpuzzleClient;
 
+    @Autowired
+    public void CrosswordpuzzleClient(final CrosswordpuzzleClient crosswordpuzzleClient) {
+        this.crosswordpuzzleClient = crosswordpuzzleClient;
+    }
+
     @Autowired
     BugfinderClient bugfinderClient;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]: ?

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 8]
```diff
@@ -453,27 +444,21 @@ private MinigameTask cloneFinitequiz(
     private MinigameTask cloneTowercrush(
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]: I am fairly certain that it is possible to refactor all these single methods into one common method…

**Review by delvh — APPROVED** [review]

**Comment by delvh** [comment]
From my side, this PR can be merged.
And let it simply be so that the CI passes again.
All other PRs only have minor changes requested, nothing serious.

**Comment by delvh** [comment]
@Aaron2000119 any complaints if I merge it now?
At this point, it simply annoys me that the CI always fails on other branches…

**Comment by Aaron2000119** [comment]
We have to change all the docker-compose-files back to the main branch otherwise it should be fine.

**Comment by delvh** [comment]
Oh, please also adapt the documentation how to integrate a new minigame once this is merged.

**Pull request merged**